File: BOJ(Java 공부용)/1774.java
import java.io.*;
import java.util.*;
import static java.lang.Math.pow;
import static java.lang.Math.sqrt;
public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        int n = Integer.parseInt(input[0]), m = Integer.parseInt(input[1]);
        ArrayList<p> v = new ArrayList<>();
        int[] parent = new int[1001];
        PriorityQueue<pp> pq = new PriorityQueue<>((a, b) -> {
            if(a.dist <b.dist) return -1;
            else if(a.dist>b.dist) return 1;
            else {
                if(a.x <b.x) return -1;
                else if(a.x >b.x) return 1;
                else {
                    if(a.y<b.y) return -1;
                    else if(a.y>b.y) return 1;
                    else return 0;
                }
            }
        });
        for(int i=0;i<=1000;i++) parent[i] = i;
        for(int i=0;i<n;i++) {
            String[] input2 = br.readLine().split(" ");
            v.add(new p(Double.parseDouble(input2[0]),Double.parseDouble(input2[1])));
        }
        int cnt = 0;
        for(int i=0;i<m;i++) {
            String[] input3 = br.readLine().split(" ");
            int a = Integer.parseInt(input3[0]), b= Integer.parseInt(input3[1]);
            if(find(a,parent)!= find(b,parent)) {
                union(a,b,parent);
                cnt++;
            }
        }

        for(int i=0;i<v.size();i++) {
            for(int j=i+1;j<v.size();j++) {
                double x = pow(v.get(i).x-v.get(j).x,2.0);
                double y = pow(v.get(i).y-v.get(j).y,2.0);
                double res = sqrt(x+y);
                pq.add(new pp(res,i+1,j+1));
            }
        }
        double sum = 0;
        while(cnt != n-1) {
            double dist = pq.peek().dist;
            int x = pq.peek().x;
            int y = pq.peek().y;
            pq.poll();
            if(find(x,parent)!=find(y,parent)) {
                union(x,y,parent);
                cnt++;
                sum += dist;
            }
        }
        System.out.println(String.format("%.2f",sum));
    }
    private static int find(int u,int[] parent) {
        if(u == parent[u]) return u;
        return parent[u] = find(parent[u],parent);
    }
    private static void union(int u,int v,int[] parent) {
        int u1 = find(u,parent),v1 = find(v,parent);
        if(u1==v1) return;
        parent[u1] = v1;
    }
    private static class pp {
        double dist;
        int x,y;
        private pp(double dist,int x,int y) {
            this.dist = dist;
            this.x = x;
            this.y = y;
        }
    }
    private static class p {
        double x,y;
        private p(double x,double y) {
            this.x = x;
            this.y = y;
        }
    }
}

File: BOJ(Python 공부용)/1774.py
import sys
from heapq import heappush, heappop
from math import sqrt

input = sys.stdin
sys.setrecursionlimit(2000)


def find(u):
    if u == parent[u]: return u
    parent[u] = find(parent[u])
    return parent[u]


def merge(u, v):
    u, v = find(u), find(v)
    if u == v: return
    parent[u] = v


if __name__ == '__main__':
    n, m = map(int, input.readline().split())
    v = []
    parent = [0] * n
    cnt, ans = 0, 0
    for i in range(n):
        parent[i] = i
    for _ in range(n):
        x, y = map(int, input.readline().split())
        v.append((x, y))

    for _ in range(m):
        a, b = map(int, input.readline().split())
        a -= 1
        b -= 1
        if find(a) != find(b):
            merge(a, b)
            cnt += 1

    pq = []
    for i in range(len(v)):
        for j in range(i + 1, len(v)):
            dist = sqrt((v[i][0] - v[j][0]) * (v[i][0] - v[j][0]) + (v[i][1] - v[j][1]) * (v[i][1] - v[j][1]))
            heappush(pq, (dist, i, j))

    while cnt != n - 1:
        dist, here, next = heappop(pq)
        if find(here) != find(next):
            merge(here, next)
            cnt += 1
            ans += dist

    print("%.2f" % ans)
